packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -379,38 +393,106 @@ subtask(TASK_COMPILE_SOLIDITY_COMPILE_JOBS)
-        .slice()
-        .sort((job1, job2) => {
-          return semver.compare(
-            job1.getSolcConfig().version,
-            job2.getSolcConfig().version
+      log(`Compiling ${compilationJobs.length} jobs`);
+
+      const versionList: string[] = [];
+      for (const job of compilationJobs) {
+        const solcVersion = job.getSolcConfig().version;
+
+        // versions older than 0.4.11 don't work with hardhat
+        // see issue https://github.com/nomiclabs/hardhat/issues/2004
+        if (semver.lt(solcVersion, COMPILE_TASK_FIRST_SOLC_VERSION_SUPPORTED)) {
+          throw new HardhatError(
+            ERRORS.BUILTIN_TASKS.COMPILE_TASK_UNSUPPORTED_SOLC_VERSION,
+            {
+              version: solcVersion,
+              firstSupportedVersion: COMPILE_TASK_FIRST_SOLC_VERSION_SUPPORTED,
+            }
           );
-        });
+        }
 
-      log(`Compiling ${sortedCompilationJobs.length} jobs`);
+        if (!versionList.includes(solcVersion)) {
+          versionList.push(solcVersion);
+        }
+      }
+
+      /**
+       * I'm not super proud of this, but we have to download the list before the compilers,
+       * otherwise the next downloads fail with a race condition.
+       *
+       * :)
+       */
+      const compilersCache = await getCompilersDir();
+      const downloader = new CompilerDownloader(compilersCache);
+      await downloader.downloadCompilersList();
+
+      /**
+       * This is a pretty hacky solution to avoid a major refactor and/or breaking changes.
+       * The hack is the fact that we're just pre-downloading all of the necessary compilers
+       * before actually running the job, so all the later code assumes the compiler is already
+       * downloaded because if it wasn't then parallel compilation will break.
+       */
+      const downloadResults = await chunkedPromiseAll(
+        versionList.map((solcVersion) => {
+          return () =>
+            run(TASK_COMPILE_SOLIDITY_GET_SOLC_BUILD, {
+              solcVersion,
+              quiet: false,
+            });
+        }),
+        os.cpus().length
+      );
 
-      const artifactsEmittedPerJob: ArtifactsEmittedPerJob = [];
-      for (let i = 0; i < sortedCompilationJobs.length; i++) {
-        const compilationJob = sortedCompilationJobs[i];
+      for (const downloadResult of downloadResults) {
+        if (HardhatError.isHardhatError(downloadResult)) {
+          throw downloadResult;
+        }
+      }
 
-        const { artifactsEmittedPerFile } = await run(
-          TASK_COMPILE_SOLIDITY_COMPILE_JOB,
-          {
-            compilationJob,
-            compilationJobs: sortedCompilationJobs,
-            compilationJobIndex: i,
-            quiet,
+      const results = await chunkedPromiseAll<
+        CompilationSuccess | CompilationFailure
+      >(
+        compilationJobs.map((compilationJob, compilationJobIndex) => {
+          return () =>
+            run(TASK_COMPILE_SOLIDITY_COMPILE_JOB, {
```

Comment:
There is a subtle problem here related to `chunkedPromiseAll` swallowing the errors. If `TASK_COMPILE_SOLIDITY_COMPILE_JOB` throws for some reason (say, the compiler process crashes or whatever), then the code will try to iterate over the error and throw an error like this:

```
An unexpected error occurred:

TypeError: errors is not iterable
```

You can check this by creating a test project and modifying the subtask to just throw some error.

Maybe we don't have to swallow the errors in `chunkedPromiseAll`? For the first usage (the downloads), that would let us remove the for-loop that checks if some result is an error. And for the second usage (the compilations), we don't throw expected errors, so we don't really need to swallow them.

This would also help simplify the code of `chunkedPromiseAll`.

Reply:
The entire point of "swallowing" the errors in `chunkedPromiseAll` is so that we can log an aggregate of the errors at the end. If I remove that logic, then it'll just throw the first error encountered like it currently does.